plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Mappings.hs

```diff
@@ -0,0 +1,174 @@
+lspTokenReverseMap = Map.fromList $ map (\x -> (toLspTokenType x, x)) $ enumFrom minBound
+
+fromLspTokenType :: SemanticTokenTypes -> Maybe HsSemanticTokenType
+fromLspTokenType tk = Map.lookup tk lspTokenReverseMap
+
+-- * 2. Mapping from GHC type and tyThing to semantic token type.
+
+
+-- | tyThingSemantic
+tyThingSemantic :: TyThing -> Maybe HsSemanticTokenType
+tyThingSemantic ty = case ty of
+  AnId vid
+    | isTyVar vid -> Just TTypeVariable
+    | isRecordSelector vid -> Just TRecField
+    | isClassOpId vid -> Just TClassMethod
+    | isFunVar vid -> Just TFunction
+    -- \| isDFunId vid -> TClassMethod
+    | otherwise -> Just TVariable
+  AConLike con -> case con of
+    RealDataCon _ -> Just TDataCon
+    PatSynCon _   -> Just TPatternSyn
+  ATyCon tyCon
+    | isTypeSynonymTyCon tyCon -> Just TTypeSyn
+    | isTypeFamilyTyCon tyCon -> Just TTypeFamily
+    | isClassTyCon tyCon -> Just TClass
+    -- fall back to TTypeCon the result
+    | otherwise -> Just TTypeCon
+  ACoAxiom _ -> Nothing
+  where
+    isFunVar :: Var -> Bool
+    isFunVar var = isFunType $ varType var
+
+isFunType :: Type -> Bool
+isFunType a = case a of
+  ForAllTy _ t -> isFunType t
+  _x           -> isFunTy a
+
+typeSemantic :: HieKind hType -> hType -> HsSemanticTokenType
+typeSemantic kind t = case kind of
+  HieFresh -> if isFunType t then TFunction else TVariable
+  HieFromDisk full_file ->
+    if isFixFunction fullType
+      then TFunction
+      else TVariable
+    where
+      fullType = recoverFullType t (hie_types full_file)
+  where
+    isFixFunction :: HieTypeFix -> Bool
+    isFixFunction (Roll x) =
+      case x of
+        HForAllTy _ a -> isFixFunction a
+        HFunTy {}     -> True
+        _             -> False
+
+-- * 3. Mapping from hieAst ContextInfo to haskell semantic token type.
+
+infoTokenType :: ContextInfo -> Maybe HsSemanticTokenType
+infoTokenType x = case x of
+  Use                      -> Nothing
+  MatchBind                -> Nothing
+  IEThing _                -> Nothing
+  TyDecl                   -> Nothing -- type signature
+  ValBind RegularBind _ _  -> Just TVariable
+  ValBind InstanceBind _ _ -> Just TClassMethod
+  PatternBind {}           -> Just TVariable
+  ClassTyDecl _            -> Just TClassMethod
+  TyVarBind _ _            -> Just TTypeVariable
+  RecField _ _             -> Just TRecField
+  -- data constructor, type constructor, type synonym, type family
+  Decl ClassDec _          -> Just TClass
+  Decl DataDec _           -> Just TTypeCon
+  Decl ConDec _            -> Just TDataCon
+  Decl SynDec _            -> Just TTypeSyn
+  Decl FamDec _            -> Just TTypeFamily
+  -- instance dec is class method
+  Decl InstDec _           -> Just TClassMethod
+  Decl PatSynDec _         -> Just TPatternSyn
+  EvidenceVarUse           -> Nothing
+  EvidenceVarBind {}       -> Nothing
+
+-- * 4. Mapping from LSP tokens to SemanticTokenOriginal.
```

Comment:
I'm still thinking about this. I wonder if we can give people a helper like this in `lsp`. It seems much better for debugging :thinking:

Reply:
I could help to open an pr to ship `recoverSemanticTokens` helper like function up to `lsp` in the future. 
I guess they should be in the `test` module of `lsp` package?